### DISCUSSION TITLE
Raspberry Pi Pico port

### DISCUSSION THREAD

**Discussion by lprader** [Moddable-OpenSource/moddable#568] [Porting]
The Moddable team has taken the first steps to get a basic port for the Raspberry Pi Pico working (see our [latest blog post](https://blog.moddable.com/blog/pico/)). We're looking forward to working with the community of developers here to finish it up.

We'll use the [`pico` branch](https://github.com/Moddable-OpenSource/moddable/tree/pico) of this repository for development and merge into the master branch when the port is complete. We can use this discussion thread to track our progress. Please feel free to jump in to the conversation and contribute to the effort!

To-do list

- [x] Builds on macOS
- [x] Builds on Linux
- [ ] Builds on Windows  (@wilberforce is [making progress](https://github.com/Moddable-OpenSource/moddable/discussions/570))
- [x] Setup instructions documented for macOS
- [x] Setup instructions documented for Linux
- [ ] Setup instructions documented for Windows
- [x] Supports standard JavaScript
- [x] Source level JavaScript debugging with `xsbug`, including runtime instrumentation
- [x] [Preloading feature](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/xs/preload.md)  of XS
- [x] [Strip feature](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/xs/XS%20Differences.md#strip-unused-features) of XS
- [x] Timer class
- [x] Cryptographic classes
- [x] Digital output 
- [x] Digital input
- [x] Digital monitor ([commit](https://github.com/Moddable-OpenSource/moddable/commit/caf1ad698be7f61ff9a69d8bfb32908b7206dbcb))
- [x] Analog input
- [x] PWM output
- [x] I²C
- [x] SPI output (needed for displays)
- [x] Audio (no example in Pico SDK?)
- [x] Look at how to integrate PIO (RMT on ESP32 might be realistic starting point)
- [ ] Atomics in XS (not currently linking)
- [ ] Web Workers (to make use of the second CPU core)
- [x] Flash read/write (to enable preferences and mods)
- [x] littlefs file system
- [ ] [Mods](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/xs/mods.md)
- [ ] Implementing TC53 IO — ECMAScript Modules for Embedded Systems    
- [x] Eliminate need to always unplug Pico and push button to enter programming mode ([Pico commit](https://github.com/Moddable-OpenSource/moddable/commit/d4830d79ef8f96db9c883ff375edde01afafe2bb), [build](https://github.com/Moddable-OpenSource/moddable/commit/7e11db645820669ff340db86707ef508978e4816) [commits](https://github.com/Moddable-OpenSource/moddable/commit/f1d2a9612ca1cdac744f434b91bc59804fef3b83))
- [x] Serial interrupt support for debugger connection (lower latency debugging) ([commit](https://github.com/Moddable-OpenSource/moddable/commit/a08a7d5269a56982cf75bd7e2de3795bfc23324b))
- [ ] Bulletproofing build / install on macOS, Windows, and Linux
- [ ] Improving setup documentation
- [ ] Testing

**Comment by microhobby** [comment]
Where can I find the "Setup instructions documented for Linux"?

**Reply by mkellner** [reply]
https://github.com/Moddable-OpenSource/moddable/blob/pico/documentation/devices/pico.md#linux-setup

Please checkout the latest version of the pico branch.

**Reply by wilberforce** [reply]
I have the blink example working:

https://github.com/Moddable-OpenSource/moddable/blob/ece6a84f3190d1191bf4698c656aad0755d6d5e3/examples/pins/blink/main.js

building within a VM and flashing the device, ending up with a flashing LED

**Comment by wilberforce** [comment]
This might be of interest...

It's a js emulator

https://github.com/wokwi/rp2040js

**Comment by anubhavgupta** [comment]
Any plans for adding support of Workers ?

**Reply by phoddie** [reply]
We'd like to. We're currently running direct-to-metal. The most practical path to Workers would be to switch over to FreeRTOS. Once that's done, Workers should more-or-less just work.

**Reply by ArashMotamedi** [reply]
Is this in development already? Or is there an estimated timeline? Pico W is a compelling package with a good price and performance balance; would love to unlock that second core.

**Comment by jakeg** [comment]
I'm currently using [KalumaJS](https://kalumajs.org/) to run JS on the Pico. How does your project compare to that?

**Reply by phoddie** [reply]
Thanks for asking. I'm not really comfortable providing comparisons between different projects, as they have different goals. I'll share a bit about what makes the Moddable SDK unique.

- The XS JavaScript engine provides full, modern JavaScript - the latest ES2023 standard. Our conformance to the JavaScript standard is on-par with web browsers.
- XS is optimized to maximize the amount of memory available to scripts by storing byte code and even object/class prototypes in flash.
- The XS [preload](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/xs/preload.md) feature begins execution on your computer and continues on the embedded devices. This allows projects to start executing a script's byte code within milliseconds of booting.
- XS minimizes the size of the engine by analyzing the byte code of your project scripts, deploying only the language features you use. (This is entirely configurable, of course).
- Our xsbug JavaScript debugger provides source level debugging, runtime instrumentation (memory use, garbage collections, network traffic, and more), and a [performance profiler](https://www.moddable.com/blog/profiler/).
- The Moddable SDK supports the [ECMA-419](https://419.ecma-international.org) standard allowing IO code, device drivers, sensor drivers, and more to work consistently across different silicon families.
- The Poco graphics renderer and Piu UI framework make building great looking user interfaces incredibly easy. Check out the [Piu reference docs](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/piu/piu.md) to get a sense of the power, and [our blog](https://www.moddable.com/blog/) to see lots of example projects.
- There's extensive support for network protocols, include a TLS secure socket implemented in JavaScript. This lets you connect to most any local or cloud service.
- The Moddable SDK is a project for professional development. It is used to deliver commercial products for consumer and industrial IoT. A great deal of effort has been put into make sure it is extremely reliable, using two extensive [test suites](https://www.moddable.com/blog/how-we-test/), and secure, through continuous fuzz testing for vulnerabilities.

There's much more. If there's something you'd like to know, please ask.

**Comment by ArashMotamedi** [comment]
Could you point me to documentation (or provide brief outline here) how to persist files to flash please? I've gotten a basic file `.write` and `.read` program working, but files seem to disappear when I power-cycle the board.

**Reply by phoddie** [reply]
File files, you would use the File module. That is [documented](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/files/files.md#filesystems) and there is an [example](https://github.com/Moddable-OpenSource/moddable/blob/public/examples/files/files/main.js).

However, the File module is not currently available on Pico. It should be straightforward to bring-up [using LittleFS](https://blog.moddable.com/blog/littlefs/). @mkellner – maybe you can help here?

All that said, often a file system is unnecessary. If you are just storing small pieces of data, our Preferences module may be a better choice. We use it far more often than the File module. That is supported on Pico. See the [documentation](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/files/files.md#preference) and [example](https://github.com/Moddable-OpenSource/moddable/blob/public/examples/files/preference/main.js).

**Comment by mkellner** [comment]
Hello!

In the next release, I will include support for building littlefs on pico.

In the interim, you can add three lines to `$MODDABLE/modules/files/file/manifest.json` to use it.

```
diff --git a/modules/files/file/manifest.json b/modules/files/file/manifest.json
index 403464c3e..6c4ef4ca6 100644
--- a/modules/files/file/manifest.json
+++ b/modules/files/file/manifest.json
@@ -92,6 +92,9 @@
                "nrf52": {
                        "include": "./manifest_littlefs.json"
                },
+               "pico": {
+                       "include": "./manifest_littlefs.json"
+               },
                "...": {
                        "error": "File module unsupported"
                }
```

The `$MODDABLE/examples/files/files` app will exercise the API.

**Reply by mkellner** [reply]
This will enable a small RAM based filesystem.
With the code in the next release, it will be backed to flash.

**Reply by ArashMotamedi** [reply]
Thank you @mkellner. Will keep an eye out for the next release. Unrelated to Files, is Web Worker taking advantage of Core1 on the roadmap as well?

**Reply by mkellner** [reply]
@ArashMotamedi the `workers` module is functional (ie. try with the example `$MODDABLE/examples/base/worker`) but it is not currently taking advantage of the second core on Pico.
